### PR TITLE
Allow packages to be built in isolation.

### PR DESCRIPTION
**Pull request by jeffstieler**
This PR is in support of upcoming organizational work in the core WC repo.

### Goal

The overarching goal of these changes is to improve the developer experience of working with the `packages/` dir, as well as fix some problems with our published packages (currently, they cannot be used without being transpiled).

When working on a package, you can now ~`yarn watch` inside the directory~ `lerna run start --scope @woocommerce/<package>` and avoid rebuilding ALL packages whenever you make a change.

#### Major changes

* ~Swapping `npm` for `yarn`. This was done to utilize the `workspaces` feature. It will likely get revisited and perhaps replaced with something like [`pnpm`](https://pnpm.io/).~ Reverted.
* Using `tsc` to build the packages instead of Babel. This _hopefully_ lays the foundation to migrate our existing code to TypeScript, but did come with the need to add `createElement` imports to all JSX files.
* Creation of a new (private) `js-tests` package to house common test bootstrapping for the `packages/` and `client/`.
* Migration of the `wc-admin-settings` to a "real" (private) package for easier resolution by `tsc`.

#### Areas for improvement

* It would be nice to not have to import `createElement` in all of our JSX files.
* There is a lot of duplication between the `.tsconfig*.json` files and the `package.json` scripts for building the JS. It might be possible to move this to another private package, but at a minimum that would require using JavaScript for the configs instead of JSON, and moving to shell scripts (in a package `bin` property for reuse - unless there's another way)
* The `bin/packages/build.js` and `bin/packages/watch.js` have been repurposed to only build CSS for a single package at a time. This might leave some prior potential usages broken (running `watch` with no arguments). Ultimately, this was a compromise made to speed up this work considering we're planning to revamp the styling of our packages and app in the (hopefully near) future.
* The storybook is built at the root project level, despite only the `components` and `experimental` packages having stories.

### Detailed test instructions:

- `rm -rf node_modules`
- `npm install`
- Test running scripts for a single package (`npm run start:package @woocommerce/<package>`)
- Test all scripts in `package.json` (some are covered by the CI env)
- `npm run build`
- Smoke test the entire application

### CODE REVIEW

**Comment by louwie17** [comment]
> Swapping npm for yarn. This was done to utilize the workspaces feature. It will likely get revisited and perhaps replaced with something like pnpm.

🎉 🎉

File: package.json [hunk 1]
```diff
@@ -18,70 +24,71 @@
...
-		"test": "./node_modules/jest-24.9.0/bin/jest.js --config tests/js/jest.config.json",
-		"test:debug": "node --inspect-brk ./node_modules/jest-24.9.0/bin/jest.js --config tests/js/jest.config.json --watch --runInBand --no-cache",
-		"test:e2e": "WP_VERSION=latest npm run build && npx wc-e2e docker:down && npx wc-e2e docker:up && npx wc-e2e test:e2e",
-		"test-staged": "./node_modules/jest-24.9.0/bin/jest.js --bail --config tests/js/jest.config.json --findRelatedTests",
+		"prepack": "yarn install && yarn run lint && yarn run test && cross-env WC_ADMIN_PHASE=core yarn run build",
+		"publish-packages:check": "yarn run build:packages && lerna updated",
+		"publish-packages:dev": "yarn run build:packages && lerna publish from-package --npm-tag next",
+		"publish-packages:prod": "yarn run build:packages && lerna publish from-package",
+		"prestart": "yarn run -s install-if-deps-outdated",
+		"start": "cross-env WC_ADMIN_PHASE=development yarn run build:packages && cross-env WC_ADMIN_PHASE=development yarn run build:feature-config && concurrently \"cross-env WC_ADMIN_PHASE=development webpack --watch\" \"yarn node ./bin/packages/watch.js\"",
```
Comment (psealock) [inline package.json:64]: I'm getting an error on this command:

```
"yarn node ./bin/packages/watch.js\"
```

Screen Shot 2021-07-06 at 12 13 14 PM

As a result, `yarn start` works but it doesn't watch the packages. It only builds them once. If I `cd packages/components` and `yarn start` at the same it, that works.

Comment (psealock) [reply package.json:64]: `yarn node ./bin/packages/watch.js packages/components` doesn't pick up changes either. 

I think `bin/packages/watch.js` could handle building all the packages if no path is given. It appears to only be used by the components package for building CSS, but thats not working.

**Review by psealock — COMMENTED** [review]
This is great work @jeffstieler 🥇 

Here is some feedback so far. I made a comment on loose mode, and building of packages isn't quite right when running `yarn start`.

Also, I was under the impression that we'd have the monorepo pull its packages from NPM instead of the local builds. This requires `yarn link` to develop locally, which is a pain, but the advantage is we are a consumer of the packages just like everyone else. Did I understand that correctly @ObliviousHarmony ?

**Comment by adrianduffell** [comment]
Wow, awesome work on this @jeffstieler! This is just some general feedback and questions mostly to help my understanding. 

> When working on a package, you can now yarn watch inside the directory and avoid rebuilding ALL packages whenever you make a change.

🥳 

> Files changed 328  +172,201 −42,060

This PR is rather huge and makes it hard to navigate and review. GitHub won’t render the Jump To dropdown for example. 

Screen Shot 2021-07-07 at 1 41 19 pm

It could perhaps be split up into separate PRs for each of the major changes you listed. There is also quite a bit of noise with random formatting changes. I’d say there’s a high risk of unknown errors creeping in here due to the PR size.

> Swapping npm for yarn. This was done to utilize the workspaces feature. It will likely get revisited and perhaps replaced with something like pnpm.

What was the reason not to just move straight to pnpm now?

What does this mean for the rest of Woo? Should other projects move away from npm as well?

Also it looks like this uses the older Yarn 1. Was there a reason not to use Yarn 2?

> Using tsc to build the packages instead of Babel. This hopefully lays the foundation to migrate our existing code to TypeScript, but did come with the need to add createElement imports to all JSX files.

I think it’s good to use `tsc`, but could you elaborate on the need to switch? What were the pain points with migrating our existing code with Babel in place?

**Comment by psealock** [comment]
@jeffstieler I confirmed `yarn test:watch` is indeed working as its supposed to, or at least matches the behaviour found in `main`. I pushed https://github.com/woocommerce/woocommerce-admin/pull/7286/commits/0c74a7a8b8a503a5e0f180a961010ed5743f684e which uses `tsc` to build packages before running tests.

**Comment by becdetat** [comment]
This is an awesome looking PR. I agree that it's a lot to review, especially with the large white-space changes to some files. It's also unfortunate that `createElement` needs to be imported everywhere, but swapping Babel out for something simpler is a very good thing IMO.

> > Swapping npm for yarn. This was done to utilize the workspaces feature. It will likely get revisited and perhaps replaced with something like pnpm.

> What was the reason not to just move straight to pnpm now?
> What does this mean for the rest of Woo? Should other projects move away from npm as well?
> Also it looks like this uses the older Yarn 1. Was there a reason not to use Yarn 2?

@jeffstieler could we have this discussion as a P2 please? I had never heard of pnpm, it's difficult to type when you're used to npm but it looks very compelling.

**Comment by jeffstieler** [comment]
@adrianduffell @becdetat 

> This PR is rather huge and makes it hard to navigate and review. GitHub won’t render the Jump To dropdown for example.
> It could perhaps be split up into separate PRs for each of the major changes you listed.

I'm working to revert the switch to `yarn`, so hopefully that will cut down on the noise ([branch here](https://github.com/woocommerce/woocommerce-admin/tree/try/back-to-npm)). I'll also see where whitespace might be muddying the waters.

> What was the reason not to just move straight to pnpm now?

It was my impression that we needed at least yarn's workspace functionality and that was a non-controversial change.

> What does this mean for the rest of Woo? Should other projects move away from npm as well?

Yes, probably, but not right this second. Other big projects at a8c are using either yarn or pnpm, like Calypso and Jetpack.

> Also it looks like this uses the older Yarn 1. Was there a reason not to use Yarn 2?

Just following the prior art at the company.

> could we have this discussion as a P2 please? I had never heard of pnpm, it's difficult to type when you're used to npm but it looks very compelling.

Absolutely. I think reverting the `yarn` changes now make sense, but ultimately the repo organization work is going to introduce a new package manager and workflow.

> I think it’s good to use `tsc`, but could you elaborate on the need to switch? What were the pain points with migrating our existing code with Babel in place?

We wanted to introduce incremental builds wherever possible. The packages were all being built together using a separate babel config from the main app (`/client`), and the shipped version of the packages looks to be broken. Based on the assumption that we'll eventually migrate all of our code to TypeScript and the desire to build incrementally (and even single packages in isolation) it made the most sent to try and use `tsc` directly.

**Comment by adrianduffell** [comment]
@jeffstieler Sounds great, thanks its much appreciated.

**Comment by jeffstieler** [comment]
@psealock I was able to revert the babel changes after ditching yarn.

**Review by psealock — COMMENTED** [review]
Nice progress here @jeffstieler 🎉 

I tested most scripts, verified the watch functionality, tests, and builds (core, dev, plugin). Its all working as it should.

`lerna --scope @woocommerce/<package> run start` requires installing lerna globally. What do you think about creating an npm script for this? This also only updates the `build-module/` files, not the ones under `build/`. I _think_ thats Ok considering the use case.

I thought we were stepping away from `"@woocommerce/components": "file:packages/components",` in favor of actual downloads from npm, but maybe thats for a follow up when we contemplate use of Yarn 3 or pnpm.

I saw you bumped every package version, which is great and we'll need accompanying changelogs too.

Otherwise, this is looking good to go

**Comment by jeffstieler** [comment]
@psealock thanks for the review!

> What do you think about creating an npm script for this?

Done in d3aabe3. Updated the testing instructions as well.

> This also only updates the build-module/ files, not the ones under build/. I think thats Ok considering the use case.

Yep, the webpack build will grab the modules.

> I thought we were stepping away from "@woocommerce/components": "file:packages/components", in favor of actual downloads from npm, but maybe thats for a follow up when we contemplate use of Yarn 3 or pnpm.

AFAIK that's going to require a move away from `npm`, or at least a retooling of our `lerna` usage (I'll take a look at this now).

> I saw you bumped every package version, which is great and we'll need accompanying changelogs too.

Done in d495bb6.

**Comment by jeffstieler** [comment]
> > I thought we were stepping away from "@woocommerce/components": "file:packages/components", in favor of actual downloads from npm, but maybe thats for a follow up when we contemplate use of Yarn 3 or pnpm.
> 
> AFAIK that's going to require a move away from npm, or at least a retooling of our lerna usage (I'll take a look at this now).

I was unsuccessful in getting this setup. 😕

**Review by psealock — APPROVED** [review]
>I was unsuccessful in getting this setup. 😕

We can leave this for further investigation when looking into build tools.

Thanks @jeffstieler, this is looking great and ready to merge 🚢

**Pull request merged**